lighthouse-core/audits/script-treemap-data.js

```diff
@@ -0,0 +1,262 @@
+    /** @type {RootNodeContainer[]} */
+    const rootNodeContainers = [];
+
+    let inlineScriptLength = 0;
+    for (const scriptElement of artifacts.ScriptElements) {
+      // No src means script is inline.
+      // Combine these ScriptElements so that inline scripts show up as a single root node.
+      if (!scriptElement.src) {
+        inlineScriptLength += (scriptElement.content || '').length;
+      }
+    }
+    if (inlineScriptLength) {
+      const name = artifacts.URL.finalUrl;
+      rootNodeContainers.push({
+        name,
+        node: {
+          name,
+          resourceBytes: inlineScriptLength,
+        },
+      });
+    }
+
+    const bundles = await JsBundles.request(artifacts, context);
+    const duplicationByPath = await ModuleDuplication.request(artifacts, context);
+
+    for (const scriptElement of artifacts.ScriptElements) {
+      if (!scriptElement.src) continue;
+
+      const name = scriptElement.src;
+      const bundle = bundles.find(bundle => scriptElement.src === bundle.script.src);
+      const scriptCoverages = artifacts.JsUsage[scriptElement.src];
+      if (!bundle || !scriptCoverages) {
+        rootNodeContainers.push({
+          name,
+          node: {
+            name,
+            resourceBytes: scriptElement.src.length,
+          },
+        });
+        continue;
+      }
+
+      const unusedJavascriptSummary = await UnusedJavaScriptSummary.request(
+        {url: scriptElement.src, scriptCoverages, bundle}, context);
+
+      /** @type {Node} */
+      let node;
+      if (unusedJavascriptSummary.sourcesWastedBytes) {
+        // Create nodes for each module in a bundle.
+
+        /** @type {Record<string, SourceData>} */
+        const sourcesData = {};
+        for (const source of Object.keys(bundle.sizes.files)) {
+          /** @type {SourceData} */
+          const sourceData = {
+            resourceBytes: bundle.sizes.files[source],
+            unusedBytes: unusedJavascriptSummary.sourcesWastedBytes[source],
+          };
+
+          const key = ModuleDuplication.normalizeSource(source);
+          if (duplicationByPath.has(key)) sourceData.duplicatedNormalizedModuleName = key;
+
+          sourcesData[source] = sourceData;
+        }
+
+        node = this.prepareTreemapNodes(bundle.rawMap.sourceRoot || '', sourcesData);
+      } else {
+        // There was no source map for this script, so we can only produce a single node.
+
+        node = {
+          name,
+          resourceBytes: unusedJavascriptSummary.totalBytes,
+          unusedBytes: unusedJavascriptSummary.wastedBytes,
+        };
+      }
+
+      rootNodeContainers.push({
+        name,
+        node,
+      });
+    }
+
+    return rootNodeContainers;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    /** @type {TreemapData} */
```

Comment:
Type doesn't seem necessary? `makeRootNodes` could have that as its return value, but it's also synonymous with the current `RootNodeContainer[]` anyways?